git_machete/cmd.py

```diff
@@ -339,359 +339,1027 @@ def popen_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs:
-        roots += [b]
-        print(fmt(f"Added branch `{b}` as a new root"))
-    else:
-        if not onto:
-            u = infer_upstream(cli_ctxt, b, condition=lambda x: x in managed_branches, reject_reason_message="this candidate is not a managed branch")
-            if not u:
-                raise MacheteException(f"Could not automatically infer upstream (parent) branch for `{b}`.\n"
-                                       "You can either:\n"
-                                       "1) specify the desired upstream branch with `--onto` or\n"
-                                       f"2) pass `--as-root` to attach `{b}` as a new root or\n"
-                                       "3) edit the definition file manually with `git machete edit`")
-            else:
-                msg = f"Add `{b}` onto the inferred upstream (parent) branch `{u}`?" + pretty_choices('y', 'N')
-                opt_yes_msg = f"Adding `{b}` onto the inferred upstream (parent) branch `{u}`"
-                if ask_if(cli_ctxt, msg, opt_yes_msg) in ('y', 'yes'):
-                    onto = u
-                else:
-                    return
+    def fork_point(self, b: str, use_overrides: bool) -> Optional[str]:
+        sha, containing_branch_defs = self.fork_point_and_containing_branch_defs(b, use_overrides)
+        return sha
 
-        up_branch[b] = onto
-        if onto in down_branches:
-            down_branches[onto].append(b)
-        else:
-            down_branches[onto] = [b]
-        print(fmt(f"Added branch `{b}` onto `{onto}`"))
+    def diff(self, branch: Optional[str]) -> None:
+        fp: str = self.fork_point(branch if branch else current_branch(self.cli_ctxt), use_overrides=True)
+        params = \
+            (["--stat"] if self.cli_ctxt.opt_stat else []) + \
+            [fp] + \
+            ([f"refs/heads/{branch}"] if branch else []) + \
+            ["--"]
+        run_git(self.cli_ctxt, "diff", *params)
 
-    save_definition_file()
+    def log(self, branch: str) -> None:
+        run_git(self.cli_ctxt, "log", "^" + self.fork_point(branch, use_overrides=True), f"refs/heads/{branch}")
 
 
-def annotate(b: str, words: List[str]) -> None:
-    global annotations
-    if b in annotations and words == ['']:
-        del annotations[b]
-    else:
-        annotations[b] = " ".join(words)
-    save_definition_file()
+def slidable(machete_context: MacheteClient) -> List[str]:
+    return [b for b in machete_context.managed_branches if b in machete_context.up_branch]
 
 
-def sync_annotations_to_github_prs(cli_ctxt: CommandLineContext) -> None:
-    global annotations, managed_branches
+def slidable_after(machete_context: MacheteClient, b: str) -> List[str]:
+    if b in machete_context.up_branch:
+        dbs = machete_context.down_branches.get(b)
+        if dbs and len(dbs) == 1:
+            return dbs
+    return []
+
+
+def sync_annotations_to_github_prs(machete_context: MacheteClient, cli_ctxt: CommandLineContext) -> None:
```

Comment:
Pls use `machete_client` name for the params of `MacheteClient` type

Reply:
Also! passing a `CommandLineContext` is probably redundant here? coz this can/should be extracted from `machete_context` instead